lib/egg_loader.js

```diff
@@ -0,0 +1,286 @@
+        // 使用 fs.realpathSync 来找到最终路径
+        const realpath = fs.realpathSync(eggPath);
+        if (eggPaths.indexOf(realpath) === -1) {
+          eggPaths.unshift(realpath);
+        }
+      }
+    }
+
+    return eggPaths;
+  }
+
+  // Low Level API
+
+  /**
+   * Load single file, will invork when export is function
+   *
+   * @param {String} filepath - fullpath
+   * @param {Array} arguments - pass rest arguments into the function when invork
+   * @return {Object} exports
+   * @example
+   * ```js
+   * app.loader.loadFile(path.join(app.options.baseDir, 'config/router.js'));
+   * ```
+   */
+  loadFile(filepath) {
+    if (!fs.existsSync(filepath)) {
+      return null;
+    }
+
+    const ret = loadFile(filepath);
+    // function(arg1, args, ...) {}
+    let inject = Array.prototype.slice.call(arguments, 1);
+    if (inject.length === 0) inject = [ this.app ];
+    return isFunction(ret) ? ret.apply(null, inject) : ret;
+  }
+
+  /**
+   * Get all loadUnit
+   *
+   * loadUnit is a directory that can be loaded by EggLoader, it has the same structure.
+   * loadUnit has a path and a type(app, framework, plugin).
+   *
+   * The order of the loadUnits:
+   *
+   * 1. plugin
+   * 2. framework
+   * 3. app
+   *
+   * @return {Array} loadUnits
+   */
+  getLoadUnits() {
+    if (this.dirs) {
+      return this.dirs;
+    }
+
+    const dirs = this.dirs = [];
+
+    // 插件目录，master 没有 plugin
+    if (this.orderPlugins) {
+      for (const plugin of this.orderPlugins) {
+        dirs.push({
+          path: plugin.path,
+          type: 'plugin',
+        });
+      }
+    }
+
+    // egg 框架路径
+    for (const eggPath of this.eggPaths) {
+      dirs.push({
+        path: eggPath,
+        type: 'framework',
+      });
+    }
+
+    // 应用目录
+    dirs.push({
+      path: this.options.baseDir,
+      type: 'app',
+    });
+
+    debug('Loaded dirs %j', dirs);
+    return dirs;
+  }
+
+  loadToApp(directory, field, opt) {
+    const target = this.app[field] = {};
+    opt = Object.assign({}, {
+      directory,
+      target,
+      inject: this.app,
+    }, opt);
+    new Loader(opt).load();
+  }
+
+  loadToContext(directory, field, opt) {
+    opt = Object.assign({}, {
+      directory,
+      field,
+      inject: this.app,
+    }, opt);
+    new ContextLoader(opt).load();
+  }
+
+}
+
+/**
+ * Mixin loader 方法到 BaseLoader，class 不支持多类继承
+ * // ES6 Multiple Inheritance
+ * https://medium.com/@leocavalcante/es6-multiple-inheritance-73a3c66d2b6b
+ */
+const loaders = [
+  require('./mixin/plugin'),
+  require('./mixin/config'),
+  require('./mixin/extend'),
+  require('./mixin/custom'),
+  require('./mixin/proxy'),
+  require('./mixin/service'),
+  require('./mixin/middleware'),
+  require('./mixin/controller'),
+];
+
+for (const loader of loaders) {
+  Object.assign(EggLoader.prototype, loader);
+}
+
+module.exports = EggLoader;
+
+function isKoa(app) {
+  return app.hasOwnProperty('use') &&
```

Comment:
这个改成 egg-core 的时候会去掉

fengmk2 <notifications@github.com>于2016年8月4日 周四上午9:07写道：

> In lib/egg_loader.js
> <https://github.com/eggjs/egg-loader/pull/5#discussion_r73448752>:
>
> > +  require('./mixin/extend'),
> > +  require('./mixin/custom'),
> > +  require('./mixin/proxy'),
> > +  require('./mixin/service'),
> > +  require('./mixin/middleware'),
> > +  require('./mixin/controller'),
> > +];
> > +
> > +for (const loader of loaders) {
> > +  Object.assign(EggLoader.prototype, loader);
> > +}
> > +
> > +module.exports = EggLoader;
> > +
> > +function isKoa(app) {
> > +  return app.hasOwnProperty('use') &&
>
> koa2 也一样？
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eggjs/egg-loader/pull/5/files/98f7bcd5a5cae8e6cddbbf037796c850ca4db42d#r73448752>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWA1VRaEC4UMDN35lFoYYrj4XAciyNwks5qcTtIgaJpZM4Jabuw>
> .
>